plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -590,25 +595,118 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        ObjectMetadata metadata = getS3ObjectMetadata(path);
+        if (metadata == null) {
+            // the bucket root does not exist
+            if (path.getName().isEmpty()) {
+                return false;
+            }
+
+            // check if this path is a directory
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(1), false);
```

Comment:
It seems like ideally we would avoid redundant listings here vs inside of `deleteObjects`